File: BOJ/14501.cpp
#include <iostream>
#include <algorithm>
#define num 17
using namespace std;
int n, scheduled[num][2], dp[num];


void input(){
    cin >> n;
    for(int i=1; i<=n; i++) {
        cin >> scheduled[i][0] >> scheduled[i][1]; //duration, payment
    }
}

void solution(){
    int answer = 0;
    for(int i=1; i<=n; i++){
        for(int j=0; j<i; j++){     //숫자의 합 i는 i 보다 작은 값으로 이루어지므로 (j: 1, 2, ... , i-1)
            if(scheduled[j][0] + j <= i && scheduled[i][0] + i <= n+1)
                dp[i] = max(dp[i], dp[j] + scheduled[i][1]);
        }
    }
    for(int i=1; i<=n; i++) answer = max(dp[i], answer);
    cout << answer << '\n';
}

int main(){
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    input();
    solution();
    return 0;
}

File: BOJ/Swift/14501.swift
import Foundation

let n = Int(readLine()!)!
var array = [(0,0)] //기간, 금액
var answer = 0

for _ in 0..<n {
    let input = readLine()!.split(separator: " ").map{ Int(String($0))! }
    array.append((input[0], input[1]))
}

func DFS(_ x: Int, _ sum: Int){
    if x == n || x + array[x].0 > n {
        answer = max(answer, sum)
        return
    }
    
    for i in array[x].0 + x...n { //다음 출근 날(인덱스)
        if i + array[i].0 - 1 > n {
            continue
        }
        DFS(i, sum + array[i].1)
    }
    
    answer = max(answer, sum)
    return
}

for d in 1...n { //시작하는 날짜
    if d + array[d].0 - 1 > n { continue }
    DFS(d, array[d].1)
}

print(answer)
